File: Client/tecmo-tourney/src/app/features/tournaments/bracket/double-elim-bracket.types.ts
import { GameStatus } from 'src/app/enums';

export type FeedKind = 'seed' | 'bye' | 'winner' | 'loser' | 'empty';

export interface FeedRef {
  kind: FeedKind;
  /** WB/LB match id, e.g. wb-0-1 or lb-2-0 */
  matchId?: string;
  /** 0-based seed index in ordered entrant list (best seed = 0) */
  seedSlot?: number;
}

export interface BracketMatch {
  id: string;
  segment: 'WB' | 'LB' | 'GF';
  /** Round within segment (GF: 0 = first final, 1 = reset) */
  round: number;
  indexInRound: number;
  top: FeedRef;
  bottom: FeedRef;
}

export interface BracketParticipant {
  playerId: number;
  name: string;
  seed: number;
}

export interface ResolvedSlot {
  participant: BracketParticipant | null;
  /** Bye placeholder */
  isBye?: boolean;
}

export interface ResolvedMatch {
  def: BracketMatch;
  top: ResolvedSlot;
  bottom: ResolvedSlot;
  gameResultId: number | null;
  status: GameStatus | null;
  topScore: number | null;
  bottomScore: number | null;
  /** Predicted winner playerId for Completed games */
  winnerId: number | null;
  /** Optional: both slots known but no game row yet */
  isPending: boolean;
  /**
   * Losers bracket only: winners-bracket game # whose loser feeds this slot (e.g. "WB2").
   * Numbering is WB1..WBn in bracket order — round 1 left→right, then round 2, etc.
   */
  topSourceLabel?: string | null;
  bottomSourceLabel?: string | null;
  /** Winners bracket: single label for the matchup (same WB# as losers bracket references). */
  wbMatchLabel?: string | null;
}

File: Service/Bracket/DoubleElimBracketTypes.cs
using TecmoTourney;

namespace TecmoTourney.Bracket;

public enum FeedKind
{
    Seed,
    Bye,
    Winner,
    Loser,
    Empty
}

public sealed class FeedRef
{
    public FeedKind Kind { get; set; }
    public string? MatchId { get; set; }
    public int? SeedSlot { get; set; }
}

public sealed class BracketMatch
{
    public string Id { get; set; } = string.Empty;
    public string Segment { get; set; } = string.Empty; // WB | LB | GF
    public int Round { get; set; }
    public int IndexInRound { get; set; }
    public FeedRef Top { get; set; } = new();
    public FeedRef Bottom { get; set; } = new();
}

public sealed class BracketParticipant
{
    public int PlayerId { get; set; }
    public string Name { get; set; } = string.Empty;
    public int Seed { get; set; }
}

public sealed class ResolvedSlot
{
    public BracketParticipant? Participant { get; set; }
    public bool IsBye { get; set; }
}

public sealed class ResolvedMatch
{
    public BracketMatch Def { get; set; } = new();
    public ResolvedSlot Top { get; set; } = new();
    public ResolvedSlot Bottom { get; set; } = new();
    public int? GameResultId { get; set; }
    public GameStatus? Status { get; set; }
    public int? TopScore { get; set; }
    public int? BottomScore { get; set; }
    public int? WinnerId { get; set; }
    public bool IsPending { get; set; }
    public string? TopSourceLabel { get; set; }
    public string? BottomSourceLabel { get; set; }
    public string? WbMatchLabel { get; set; }
}

/// <summary>Minimal game facts for bracket resolution (mirrors Angular IGameResult fields used in resolve).</summary>
public sealed class BracketGameSnapshot
{
    public int GameResultId { get; set; }
    public int Player1Id { get; set; }
    public int Player2Id { get; set; }
    public int Player1Score { get; set; }
    public int Player2Score { get; set; }
    public GameStatus Status { get; set; }
    public DateTime DateAdded { get; set; }
}
